src/crimson/osd/replicated_backend.cc

```diff
@@ -38,31 +40,337 @@ ReplicatedBackend::_read(const hobject_t& hoid,
+    return get_target_oid(op) == parent_idx;
+  });
+  f(std::move(rng));
+}
+
+
+// This piece takes `os::Transaction` and scans it for multi-object
+// operations to call the passed lambda in the same order as in OSD
+// is happens thankfully to `PGTransaction::safe_create_traverse()`.
+template <class F>
+static void safe_create_traverse(crimson::osd::HLTransaction& txn, F&& f)
+{
+  for (const auto& [obj, idx] : txn.get_object_index()) {
+    // the `is_source_for` releationship between the operations in txn
+    // forms a graph. we're interested only in forrest of trees and we
+    // process these trees from their roots. if the graph contains any
+    // cycle, this selection drops it.
+    logger().debug("{}: obj={} idx={}", __func__, obj, idx);
+    if (!is_source_for_any(idx, txn.get_object_index(), txn)) {
+      logger().debug("{}: obj={} is root!", __func__, obj);
+      process_tree(txn, idx, f);
+    }
+  }
+}
+
+template <class TxnT>
+static std::string dump_txn(TxnT& txn)
+{
+  std::unique_ptr<Formatter> f{Formatter::create("json-pretty", "json-pretty", "json-pretty")};
+  std::stringstream ss;
+  try {
+    txn.dump(f.get());
+  } catch (const ceph::buffer::end_of_buffer& e) {
+    ss << "GOT EXCEPTION! e=" << e << std::endl;
+  }
+  f->flush(ss);
+  return ss.str();
+}
+
 ReplicatedBackend::rep_op_fut_t
 ReplicatedBackend::_submit_transaction(std::set<pg_shard_t>&& pg_shards,
                                        const hobject_t& hoid,
-                                       ceph::os::Transaction&& txn,
+                                       ceph::os::Transaction&& _txn,
                                        osd_op_params_t&& osd_op_p,
                                        epoch_t min_epoch, epoch_t map_epoch,
 				       std::vector<pg_log_entry_t>&& log_entries)
 {
+  crimson::osd::HLTransaction hltxn(std::move(_txn));
+  logger().debug("{}: txn={}", __func__, dump_txn(hltxn));
   if (__builtin_expect(stopping, false)) {
     throw crimson::common::system_shutdown_exception();
   }
   if (__builtin_expect((bool)peering, false)) {
     throw crimson::common::actingset_changed(peering->is_primary);
   }
 
+  safe_create_traverse(hltxn, [&hltxn](auto&& rng) {
+    logger().debug("_submit_transaction: subrange");
+#if 0
+    std::ranges::stable_sort(rng, [](const auto& lhs, const auto& rhs) {
+      if (lhs.op == ceph::os::Transaction::OP_CREATE) {
+        if (rhs.op != ceph::os::Transaction::OP_CREATE) {
+          return true;
+        }
+      } else if (lhs.op == ceph::os::Transaction::OP_CLONE) {
+        if (rhs.op == ceph::os::Transaction::OP_CREATE) {
+          return false;
+        } else if (rhs.op == ceph::os::Transaction::OP_CLONE) {
+          return false;
+        } else {
+          return true;
+        }
+      }
+      return false;
+    });
+#else
+    std::ranges::for_each(std::move(rng), [&hltxn](auto& op) {
+      hltxn.reorder_back(op);
+    });
+#endif
```

Comment:
?

Reply:
`PGTransaction` causes reordering to happen not only between the nodes but also within them (create-or-clone, then writes and so on). While testing in crimson-rados I haven't encountered the need for this yet.